src/tcache.c

```diff
@@ -660,17 +676,19 @@ tcache_init(tsd_t *tsd, tcache_slow_t *tcache_slow, tcache_t *tcache,
 
 	/*
 	 * We reserve cache bins for all small size classes, even if some may
-	 * not get used (i.e. bins higher than nhbins).  This allows the fast
-	 * and common paths to access cache bin metadata safely w/o worrying
-	 * about which ones are disabled.
+	 * not get used (i.e. bins higher than thread_nhbins).  This allows
+	 * the fast and common paths to access cache bin metadata safely w/o
+	 * worrying about which ones are disabled.
 	 */
-	unsigned n_reserved_bins = nhbins < SC_NBINS ? SC_NBINS : nhbins;
+	unsigned thread_nhbins = thread_nhbins_get(tsd);
```

Comment:
This function is not only for auto tcache in tsd. Let's use the nhbins / tcache_max from the input tcache (and avoid the thread prefix in the var names), instead of bypassing that and load from tsd.

Another request, can you please check all the references of thread_nhbins / _tcache_max_get(), and see if they can be replaced by the non-thread variant? Most of them should have tcache in their context, and access the values from the tcache. This should help when we add support for manual tcache.

This made me wonder, maybe providing the thread_nhbins_ / _tcache_max wrappers is a mistake -- we should always access it from the tcache, and the convenience from thread_nhbins_get means it's very easy to make things inconsistent. In fact I'd guess even if we provide the thread_ wrappers, there shouldn't be more than a few use cases for them. WDYT @guangli-dai ?

Reply:
Thank you for the comments! Reviewed all the usages. It's no doubt more convenient to keep a thread_nhbins/ _tcache_max_get and even eliminate them, the only difference is we explicitly get the tcache pointer and read values there repeatedly. What kinds of inconsistency concerns you?

Reply:
I see. That makes sense and it is indeed easy to forget the manual tcache scenarios with the thread_tcache_max/nhbins_get. I will go ahead and eliminate this variant. Will also modify the PR statement later, I think we should call it "Enable per-tcache tcache_max" instead.